src/main/java/org/prlprg/rds/RDSWriter.java

```diff
@@ -0,0 +1,552 @@
+import org.prlprg.RVersion;
+import org.prlprg.primitive.Logical;
+import org.prlprg.sexp.*;
+
+public class RDSWriter implements Closeable {
+  private final RDSOutputStream out;
+  // refIndex is 1-based, so the first ref will have index 1
+  private int refIndex = 1;
+  private final HashMap<SEXP, Integer> refTable = new HashMap<>(128);
+
+  protected RDSWriter(OutputStream out) {
+    this.out =
+        Boolean.parseBoolean(System.getenv("VERBOSE"))
+            ? new RDSOutputStreamVerbose(out)
+            : new RDSOutputStream(out);
+  }
+
+  public static void writeStream(RSession session, OutputStream output, SEXP sexp)
+      throws IOException {
+    try (var writer = new RDSWriter(output)) {
+      writer.write(sexp);
+    }
+  }
+
+  public static void writeFile(RSession session, File file, SEXP sexp) throws IOException {
+    try (var output = new FileOutputStream(file)) {
+      writeStream(session, output, sexp);
+    }
+  }
+
+  public void writeHeader() throws IOException {
+    // Could also be "B" (binary) and "A" (ASCII) but we only support XDR.
+    // XDR just means big endian and DataInputStream/DataOutputStream from Java use BigEndian
+    out.writeByte((byte) 'X');
+    out.writeByte((byte) '\n');
+
+    // Write version 2 of the encoding, since we want the writer to align with the reader and
+    // the reader does not support ALTREP
+    out.writeInt(2);
+
+    // Version of R for the writer
+    out.writeInt(RVersion.LATEST_AWARE.encode());
+
+    // Minimal version of R required to read back
+    out.writeInt((new RVersion(3, 5, 0, null)).encode());
+  }
+
+  public void write(SEXP sexp) throws IOException {
+    writeHeader();
+    writeItem(sexp);
+  }
+
+  // See
+  // https://github.com/wch/r-source/blob/65892cc124ac20a44950e6e432f9860b1d6e9bf4/src/main/serialize.c#L1021
+  public void writeItem(SEXP s) throws IOException {
+    var type = rdsType(s);
+
+    // Persisted through the ref table? TODO
+
+    switch (type) {
+      case RDSItemType.Special special -> {
+        // Save Special hooks: direct return and exit the function after a special hook
+        out.writeInt(special.i());
+      }
+      case RDSItemType.Sexp sexp -> {
+        // If refIndex is greater than 0, the object has already been written so we can
+        // write a reference (since the index is 1-based)
+        int sexpRefIndex = refTable.getOrDefault(s, 0);
+        if (sexpRefIndex > 0) {
+          writeRef(sexpRefIndex);
+          return;
+        }
+
+        // Otherwise, write the sexp as normal
+        switch (s) {
+          case RegSymSXP sym -> writeSymbol(sym);
+          case EnvSXP env -> writeEnv(env);
+
+            // Dotted-pair (pairlist) objects
+          case ListSXP list -> writeListSXP(list);
+          case LangSXP lang -> writeLangSXP(lang);
+          case PromSXP prom -> writePromSXP(prom);
+          case CloSXP clo -> writeCloSXP(clo);
+
+            // TODO: external pointer
```

Comment:
What about those ones?

We can do the same as R does and serialize a null pointer here (possibly with the right tag).